src/worker/core/player/shouldRetire.ts

```diff
@@ -9,13 +9,39 @@ import type {
 const shouldRetire = (
 	p: Player<MinimalPlayerRatings> | PlayerWithoutKey<MinimalPlayerRatings>,
 ): boolean => {
-	const age = g.get("season") - p.born.year;
-
+  
+	const season = g.get("season");
+	const originalSeason = g.get("startingSeason");
 	const forceRetireAge = g.get("forceRetireAge");
+	const forceRetireSeason = g.get("forceRetireSeason");
+
+	const age = season - p.born.year;
+
 	if (forceRetireAge >= g.get("draftAges")[1] && age >= forceRetireAge) {
 		return true;
 	}
 
+	//check how many unique seasons the player has played in
+	let uniqueSeasons = 0;
+	const seasonArray = new Array(season).fill(0);
+	p.stats.forEach(item => {
+		if (seasonArray[item.year] == 0) {
+			uniqueSeasons++;
+		}
+		seasonArray[item.year]++;
+	});
+
+	//if the player has played less unique seasons than the force retire number give them 1 more
+	//season of eligibility
+	let redshirt = 0;
+	if (uniqueSeasons < forceRetireSeason && originalSeason < p.draft.year) {
+		redshirt = 1;
+	}
```

Comment:
This doesn't handle the case where a player was on a team but with 0 GP (such as due to injury).

This is how I'd do it, although I haven't actually run this code so it may have a bug! General idea is to first start with the list of possible seasons, then find all the ones that are *not* redshirt seasons (GP > 0), and then use that to see if there are any redshirt seasons left at the end. I think you're right you have to iterate over all stat rows here.

NUM_REDSHIRT_YEARS_ALLOWED is just to allow some future flexibility, I imagine someone might ask for that as a separate setting later, but for now hardcoding it is fine I think.

Also feel free to disagree, don't assume my version is perfect. In fact I just rewrote it as I am writing this comment, and the previous version was uglier!

```
import range from "lodash-es/range";

const checkForceRetireSeason = (p: PlayerWithoutKey<MinimalPlayerRatings>) => {
    // Initialize with all possible seasons played, in case a player was unsigned an entire year so there is no entry in p.stats
    const redshirtSeasons = new Set(range(p.draft.year + 1, season + 1));

    // If a season has games played, it can't be a redshirt season
    for (const row of p.stats) {
        if (row.gp > 0) {
            redshirtSeasons.delete(row.season);
        }
    }

    const NUM_REDSHIRT_YEARS_ALLOWED = 1;
    const numRedshirtSeasons = Math.min(NUM_REDSHIRT_YEARS_ALLOWED, redshirtSeasons.size);
    const numSeasonsInLeague = g.get("season") - p.draft.year;

    return numSeasonsInLeague - numRedShirtSeasons >= g.get("forceRetireSeason");
}
```